enhancements/windows-containers/bring-your-own-windows-host.md

```diff
@@ -0,0 +1,182 @@
+components on Windows nodes upgraded when new Kubernetes versions or patch
+versions are released.
+
+### Risks and Mitigations
+
+The same risk and mitigation that apply to Windows nodes added using MachineSets
+to an IPI cluster applies here. In addition there is potential for a for a node
+to become unstable if configuration fails and will require administrator
+intervention to fix. There is no real mitigation here other that for us to test
+this feature thoroughly in CI.
+
+## Design Details
+
+WMCO will include a ConfigMap controller that watches for changes to the
+_windows-instances_ ConfigMap. Any change will cause WMCO to compare the list
+of instances with the set of existing Windows nodes. The comparison will be
+based on the instance and node's IP addresses. WMCO will then react in the
+following manner:
+* If a new entry is found i.e. the list of nodes does not contain one that map
+  to the instances' IP address, the instance will be configured by WMCO. The
+  configuration process will be identical to configuring instances added using
+  a MachineSet in an IPI cluster except in addition, WMCO will also have to
+  handle CSR approvals.
+* If the list of nodes has IP addresses that do not map to any in the list of
+  instances, WMCO will perform an uninstall operation on the node. This will
+  involve:
+  * Drain the node
+  * Stop and remove
+    * Monitoring and logging services
+    * kube-proxy service
+    * hybrid-overlay-node service
+    * kubelet service
+  * Delete the payload binaries
+* If the list of nodes has IP addresses that do not map to any in the list of
+  instances, and WMCO is unable to access the instance, it will assume that
+  the node in question is not managed by it.
+
+The status of these operations will be reflected by events in the
+_openshift-windows-machine-config-operator_ namespace and will be associated
+with the _windows-instances_ ConfigMap.
+
+The upgrade process for BYOH nodes will be identical to upgrading Windows nodes
+added using a MachineSet in an IPI cluster.
+
+### Test Plan
+
+The existing set of WMCO end to end tests will be run against already created
+Windows instances. This can be achieved by creating a Windows Machine that does
+not have the _machine.openshift.io/os-id: Windows_ label. Once this Machine has
+been created the Machine API and cluster provider controllers should be
+disabled to prevent them from reacting. Then the created Machine's IP can then
+be added to / removed from the _windows-instances_ ConfigMap which will cause
+WMCO to react.
+
+### Graduation Criteria
+
+This enhancement will start as GA.
+
+### Upgrade / Downgrade Strategy
+
+The upgrade process for BYOH nodes will be different to that of
+[upgrading Windows nodes](https://github.com/openshift/enhancements/blob/master/enhancements/windows-containers/windows-machine-config-operator-upgrades.md)
+added using a MachineSet in an IPI cluster. In the MachineSet case, Machines
```

Comment:
I'd like to head more about the scope of this, We need to be careful that if we are making certain changes to the node in WMCO version N, that we are either undoing them before configuring the node in WMCO version N+1. I'm thinking of things like the hybrid overlay, as it is creating HNS networks on the host which could cause complications in future versions.

Reply:
A possible solution for this is creating a configmap within the `openshift-windows-machine-config-operator` namespace, which will be used for keep track of files and services created by previous versions of the WMCO. When a newer version of the WMCO upgrades the node, it will use the configmap to determine what needs to be cleaned up.

So if the configmap has the contents:
```
files:
  c:/k
 services:
   kubelet
   windows-exporter
```
After the node has been drained the services `kubelet` and `windows-exporter` will be removed, and `c:\k` will be removed. 
The contents of the configmap can be cumulative as the WMCO is upgraded.

This approach solves an issue of having to maintain specific cleanup code for previous versions of the operator within new versions of the operator.

Reply:
I think this is outside the scope of the BYOH proposal. We need to do this in general for both Machines and BYOH nodes in a holistic manner and suggest doing this as a separate enhancement. Moreover where we stand today, we don't have 10s of services but just a handful of services. Given that the amount of configuration for services in the code is minimal.